flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java

```diff
@@ -0,0 +1,372 @@
+		}
+		resourceManager.close();
+	}
+
+	class TestingKubernetesResourceManager extends KubernetesResourceManager {
+
+		private SlotManager slotManager;
+
+		public TestingKubernetesResourceManager(
+				RpcService rpcService,
+				String resourceManagerEndpointId,
+				ResourceID resourceId,
+				Configuration flinkConfig,
+				HighAvailabilityServices highAvailabilityServices,
+				HeartbeatServices heartbeatServices,
+				SlotManager slotManager,
+				JobLeaderIdService jobLeaderIdService,
+				ClusterInformation clusterInformation,
+				FatalErrorHandler fatalErrorHandler,
+				ResourceManagerMetricGroup resourceManagerMetricGroup) {
+			super(
+				rpcService,
+				resourceManagerEndpointId,
+				resourceId,
+				flinkConfig,
+				highAvailabilityServices,
+				heartbeatServices,
+				slotManager,
+				jobLeaderIdService,
+				clusterInformation,
+				fatalErrorHandler,
+				resourceManagerMetricGroup
+			);
+			this.slotManager = slotManager;
+		}
+
+		@Override
+		protected FlinkKubeClient createFlinkKubeClient() {
+			return flinkKubeClient;
+		}
+
+		<T> CompletableFuture<T> runInMainThread(Callable<T> callable) {
+			return callAsync(callable, TIMEOUT);
+		}
+
+		@Override
+		protected void runAsync(Runnable runnable) {
+			runnable.run();
+		}
+
+		MainThreadExecutor getMainThreadExecutorForTesting() {
+			return super.getMainThreadExecutor();
+		}
+
+		SlotManager getSlotManager() {
+			return this.slotManager;
+		}
+	}
+
+	@Test
+	public void testStartAndStopWorker() throws Exception {
+		registerSlotRequest(resourceManager);
+
+		final KubernetesClient client = getKubeClient();
+		final PodList list = client.pods().list();
+		assertEquals(1, list.getItems().size());
+		final Pod pod = list.getItems().get(0);
+
+		final Map<String, String> labels = getCommonLabels();
+		labels.put(Constants.LABEL_COMPONENT_KEY, Constants.LABEL_COMPONENT_TASK_MANAGER);
+		assertEquals(labels, pod.getMetadata().getLabels());
+
+		assertEquals(1, pod.getSpec().getContainers().size());
+		final Container tmContainer = pod.getSpec().getContainers().get(0);
+		assertEquals(CONTAINER_IMAGE, tmContainer.getImage());
+
+		final String podName = CLUSTER_ID + "-taskmanager-1-1";
+		assertEquals(podName, pod.getMetadata().getName());
+
+		// Check environments
+		assertThat(tmContainer.getEnv(), Matchers.contains(
+			new EnvVarBuilder().withName(Constants.ENV_FLINK_POD_NAME).withValue(podName).build()));
+
+		// Check task manager main class args.
+		assertEquals(3, tmContainer.getArgs().size());
+		final String confDirOption = "--configDir " + flinkConfig.getString(KubernetesConfigOptions.FLINK_CONF_DIR);
+		assertTrue(tmContainer.getArgs().get(2).contains(confDirOption));
+
+		resourceManager.onAdded(Collections.singletonList(new KubernetesPod(flinkConfig, pod)));
+		final ResourceID resourceID = new ResourceID(podName);
+		assertThat(resourceManager.getWorkerNodes().keySet(), Matchers.contains(resourceID));
+
+		registerTaskExecutor(resourceID);
```

Comment:
Actually we do not need rely on the complicated TM registration/unregistration logics to verify the internal implementations inside `KubernetesResourceManager`. We can call `KubernetesResourceManager#stopWorker` to verify the function directly. Any interactions between `ResourceManager#registerTaskExecutor` and `ResourceManager#stopWorker` should already be covered by history codes. And we would also have an it case finally to verify the whole Kubernetes story.

Another benefit to do so is that we can avoid unnecessary effecting by any registration/unregistration relevant changes future to make this test failure. E.g.  if there are any argument changes future in `ResourceManager#registerTaskExecutor` or `SlotReport/SlotStatus`, this test also needs to be adjusted accordingly.

Reply:
But i can also accept to retain the current way since we already did it to not waste efforts. Just explain my thoughts for this issue.

Reply:
Make sense to me. However, it just seems that the Yarn has the similar verification logics. So currently i will keep the `registerSlotRequest` `registerTaskExecutor` `unregisterTaskManagers` to verify the `KubernetesResourceManager#startNewWorker` `KubernetesResourceManager#stopWorker`.